sounddevice.py

```diff
@@ -2349,6 +2376,97 @@ def __init__(self, channel_selectors):
             channelSelectors=self._selectors))
 
 
+class MacCoreSettings(object):
+
+    def __init__(self, channel_map=None, change_dev_params=False, prevent_rate_conv=False, rate_conv_quality='Max'):
+        """Mac Core Audio-specific input/output settings.
+
+        Objects of this class can be used as *extra_settings* argument
+        to `Stream()` (and variants) or as `default.extra_settings`.
+
+        Parameters
+        ----------
+        channel_map : iterable of int, optional
+            Support for opening only specific channels of a MacCore
+            device.  *channel_map* is a list of integers specifying the
+            (zero-based) channel numbers to use.  Note that
+            *channel_map* is treated differently between input and
+            output channels.  See the example below.
+        change_dev_params : bool, optional
+            If ``True``, allows PortAudio to change things like the
+            device's frame size, which allows for much lower latency,
+            but might disrupt the device if other programs are using it,
+            even when you are just Querying the device.  ``False`` is
+            the default.
+        prevent_rate_conv : bool, optional
+            In combination with the above flag, ``True`` causes the
+            stream opening to fail, unless the exact sample rates are
+            supported by the device.
+        rate_conv_quality : string, optional
+            This sets MacCore's sample rate conversion quality.
+            Possible values are: ``'Min'``, ``'Low'``, ``'Medium'``,
+            ``'High'``, or ``'Max'``.  ``'Max'`` is the default.
+
+        Example
+        -------
+        Input from 'Mic 9' and 'Mic 10' on RME Fireface UFX+, and send
+        output to 'Analog 3' and 'Analog 4':
+
+        >>> import sounddevice as sd
+        >>>
+        >>> input_channels = [8, 9]
+        >>> output_channels = [2, 3]
+        >>>
+        >>> # For input devices, the channel_map is simply the list of
+        >>> # 0-based index numbers for each channel:
+        >>> ch_map_in = input_channels
+        >>> mc_in = sd.MacCoreSettings(channel_map=ch_map_in)
+        >>>
+        >>> # For output devices, the list needs to be "inverted":
+        >>> # Populate the array with -1 for unwanted channels:
+        >>> max_output_channels = query_devices(device, 'output')['max_output_channels']
+        >>> ch_map_out = [-1 for _ in range(max_output_channels)]
+        >>> # For this example device, max_output_channels == 94.
+        >>> # Now overwrite any -1's for channels we'll use for output:
+        >>> for ndx in range(len(output_channels)):
+        >>>     ch_map_out[output_channels[ndx]]=ndx
```

Comment:
I didn't find much information on the matter either.  The closest notes I found were also at https://app.assembla.com/spaces/portaudio/git/source/master/src/hostapi/coreaudio/notes.txt which does state to include the extra -1's for unused channels at the end.

I'm wary to cut it any shorter if it accidentally works today, but isn't guaranteed to continue working in the future.  And the memory cost of a few dozen or even hundred int's isn't too much, considering the memory may be freed after the open is finished, right?

But all this complexity with the -1 for unused channels is what drove my initial hunch to include this array book-keeping on behalf of the caller.  &lt;shrug/&gt;  I'm still fine with leaving it out though at the expense of more words in the doc string and more platform-specific annoyances for those lost souls (like myself...) on Core Audio systems.  :)

Anyhow, since the current MacCoreSettings/CoreAudioSettings class punts all responsibility for array formatting to the caller, it can be left up to application code on how paranoid (or not) they want to be.